### PR TITLE
feat: Adds plugin-chart-handlebars

### PR DESCRIPTION
**Pull request by jdbranham**
Adds a new plugin, that renders the data payload
by applying a customizable handlebars template  

Reopening PR from old repo -  
https://github.com/apache-superset/superset-ui/pull/1390 

💔 Breaking Changes - None

🏆 Enhancements - New Handlebars plugin

📜 Documentation - slim 😅

Markup -  
markup

Markdown -  
markdown

### CODE REVIEW

**Comment by mayurnewase** [comment]
This is so cool..
Approved workflows.

**Comment by villebro** [comment]
Thanks @jdbranham so much for bringing in this PR from `superset-ui` to the main repo! Tagging myself for a review + will be reviewing+testing this in the coming days.

File: superset-frontend/plugins/plugin-chart-handlebars/src/components/Handlebars/HandlebarsViewer.tsx
Comment (villebro) [inline superset-frontend/plugins/plugin-chart-handlebars/src/components/Handlebars/HandlebarsViewer.tsx:21]: I would add a try-catch block here to catch compilation errors. Something like
```typescript
const [error, setError] = useState('');

  useMemo(() => {
    try {
      const template = Handlebars.compile(templateSource);
      const result = template(data);
      setRenderedTemplate(result);
      setError('');
    } catch (error) {
      setRenderedTemplate('');
      setError(error.message);
    }
  }, [templateSource, data]);
```

You could then add an Error component that displays the error, like

```typescript
const Error = styled.pre`
  white-space: pre-wrap;
`;

  if (error) {
    return <Error>{error}</Error>;
  }
```

This would produce something like this when the template has a syntax error:

image

File: superset-frontend/plugins/plugin-chart-handlebars/src/components/Handlebars/HandlebarsViewer.tsx
Comment (villebro) [inline superset-frontend/plugins/plugin-chart-handlebars/src/components/Handlebars/HandlebarsViewer.tsx:3]: nit: it appears `moment` isn't listed as a dependency. Since `moment` is a dependency in the monorepo, we can probably add it as a peer dependency in `package.json`

File: superset-frontend/plugins/plugin-chart-handlebars/src/plugin/controls/handlebarTemplate.tsx
Comment (villebro) [inline superset-frontend/plugins/plugin-chart-handlebars/src/plugin/controls/handlebarTemplate.tsx:59]: nit: maybe we could just inline the `handlebarsTemplateControlConfig ` config from above here?

File: superset-frontend/plugins/plugin-chart-handlebars/src/plugin/controls/handlebarTemplate.tsx
Comment (villebro) [inline superset-frontend/plugins/plugin-chart-handlebars/src/plugin/controls/handlebarTemplate.tsx:53]: let's add
```typescript
{
  ...,
  renderTrigger: true,
}
```
here to avoid having to retrigger the data request every time the template is updated (this make the chart update in real time when updating the template 🎉 )

File: superset-frontend/plugins/plugin-chart-handlebars/src/plugin/controls/handlebarTemplate.tsx
Comment (villebro) [inline superset-frontend/plugins/plugin-chart-handlebars/src/plugin/controls/handlebarTemplate.tsx:57]: nit, I would probably make this camelCase instead of PascalCase: 
```suggestion
export const handlebarsTemplateControlSetItem: ControlSetItem = {
```

File: superset-frontend/plugins/plugin-chart-handlebars/src/plugin/controls/style.tsx
Comment (villebro) [inline superset-frontend/plugins/plugin-chart-handlebars/src/plugin/controls/style.tsx:54]: nit: also inline config here

File: superset-frontend/plugins/plugin-chart-handlebars/src/plugin/controls/style.tsx
Comment (villebro) [inline superset-frontend/plugins/plugin-chart-handlebars/src/plugin/controls/style.tsx:48]: also add `renderTrigger: true` here

**Review by villebro — COMMENTED** [review]
@jdbranham I absolutely ❤️LOVE❤️ this plugin! First review pass along with a few comments - if you wish I can push some commits to the PR, e.g. resolve the conflicts and apply those proposed changes. Let me know how you want to proceed!

File: superset-frontend/plugins/plugin-chart-handlebars/src/components/Handlebars/HandlebarsViewer.tsx
Comment (jdbranham) [reply superset-frontend/plugins/plugin-chart-handlebars/src/components/Handlebars/HandlebarsViewer.tsx:21]: Excellent suggestion!

File: superset-frontend/plugins/plugin-chart-handlebars/src/plugin/controls/handlebarTemplate.tsx
Comment (jdbranham) [reply superset-frontend/plugins/plugin-chart-handlebars/src/plugin/controls/handlebarTemplate.tsx:39]: If we make it camelCase, it would be different then the other ControlSetItems 🤔 

https://github.com/apache/superset/pull/17903/files#diff-c424dd95a62b793c9f111b432e365b8cb478f2320675a7d76c05afc917c109dfR133-R153

Comment (villebro) [reply superset-frontend/plugins/plugin-chart-handlebars/src/plugin/controls/handlebarTemplate.tsx:39]: Yes; since they instantiate the `ControlSetItem` type, I consider naming them like regular JS variables (camelCase) more appropriate than `PascalCase`. So I would recommend changing the case on all of these, as it's the common convention throughout the other control definitions.

File: superset-frontend/plugins/plugin-chart-handlebars/src/plugin/controls/metrics.tsx
Comment (villebro) [inline superset-frontend/plugins/plugin-chart-handlebars/src/plugin/controls/metrics.tsx:40]: The convention has been to make all these camelCase, as they've really just variables (making them PascalCase makes it appear as if they're types/interfaces/classes/React components)

File: superset-frontend/plugins/plugin-chart-handlebars/src/plugin/controls/handlebarTemplate.tsx
Comment (villebro) [reply superset-frontend/plugins/plugin-chart-handlebars/src/plugin/controls/handlebarTemplate.tsx:39]: @jdbranham here is an example of instances of `ControlSetItem` in another plugin which are camelCase: https://github.com/apache/superset/blob/master/superset-frontend/plugins/plugin-chart-echarts/src/controls.tsx

**Review by villebro — COMMENTED** [review]
This has gotten conflicted again 🙁 Tested and it works really nicely now - after this is rebased and you've had the opportunity to check my last comment about camelCasing the controls then I think this should be good to go!

**Comment by raags** [comment]
@jdbranham would love to try this plugin - could you fix the conflicts?

**Comment by villebro** [comment]
@jdbranham I've fixed the conflicts and outstanding review comments, but I'm unable to push to your PR (I assume the branch is protected). I'd really love for this plugin to be merged in time for the 2.0 cut, so  please let me know if I can help get this merged.

**Comment by jdbranham** [comment]
That's great @villebro - I've been too swamped to get back to it 😓 

I should be able to check a box on this PR to allow superset maintainers the ability to push to the fork branch, but I don't see it.  
Looking for an alternative way...

**Comment by jdbranham** [comment]
@villebro  I just added you to my organization instead. That should work!

**Comment by villebro** [comment]
Awesome @jdbranham ! ❤️

**Comment by villebro** [comment]
@jdbranham hmm I'm still unable to push to the PR. But I created this temporary branch if you want to merge it into your branch: https://github.com/apache/superset/tree/jb/feat-plugin-handlebars . After that it should be good to be merged.

**Comment by raags** [comment]
Is it possible to include this in 1.5?

**Comment by jdbranham** [comment]
Thanks @villebro - we're merged up!

**Comment by villebro** [comment]
@jdbranham @raags if we can get this merged today I'm def gonna include it in 1.5!

**Comment by villebro** [comment]
@jdbranham there's a few minor lint issues, here's the diff which should fix the remaining errors:
```patch
diff --git a/superset-frontend/plugins/plugin-chart-handlebars/test/plugin/buildQuery.test.ts b/superset-frontend/plugins/plugin-chart-handlebars/test/plugin/buildQuery.test.ts
index d7c0457dfa..ec0eb1bae2 100644
--- a/superset-frontend/plugins/plugin-chart-handlebars/test/plugin/buildQuery.test.ts
+++ b/superset-frontend/plugins/plugin-chart-handlebars/test/plugin/buildQuery.test.ts
@@ -7,8 +7,8 @@ describe('Handlebars buildQuery', () => {
     granularitySqla: 'ds',
     groupby: ['foo'],
     viz_type: 'my_chart',
-    width: '500px',
-    height: '500px'
+    width: 500,
+    height: 500,
   };

   it('should build groupby with series in form data', () => {
diff --git a/superset-frontend/plugins/plugin-chart-handlebars/test/plugin/transformProps.test.ts b/superset-frontend/plugins/plugin-chart-handlebars/test/plugin/transformProps.test.ts
index f24d2c29f4..b5dc643529 100644
--- a/superset-frontend/plugins/plugin-chart-handlebars/test/plugin/transformProps.test.ts
+++ b/superset-frontend/plugins/plugin-chart-handlebars/test/plugin/transformProps.test.ts
@@ -9,8 +9,9 @@ describe('Handlebars tranformProps', () => {
     granularitySqla: 'ds',
     metric: 'sum__num',
     groupby: ['name'],
-    width: '500px',
-    height: '500px'
+    width: 500,
+    height: 500,
+    viz_type: 'handlebars',
   };
   const chartProps = new ChartProps<QueryFormData>({
     formData,
```

**Comment by villebro** [comment]
@jdbranham I'm probably going to cut 1.5.0rc4 in ~20 hours, so if you can push this last change in before then, I'll include this in 1.5.0 🚀

**Comment by jdbranham** [comment]
hope I made it in time!

**Comment by villebro** [comment]
Thanks @jdbranham! Fingers crossed! 🤞

**Comment by villebro** [comment]
@jdbranham oh no, those widths and heights needed to be numbers (not strings), so the linter is still failing 😢

**Comment by jdbranham** [comment]
also seeing this 🤔 
image

**Comment by jdbranham** [comment]
It's strange because when I run the tests locally the result is different than GHA 😢

**Comment by jdbranham** [comment]
I must have a problem with my environment. The tests in GHA show the exact opposite when I run locally.
Unless I'm comparing the wrong runs... I'll double check.

When I remove `viz_type`
image


And when I add it back - 
image

**Comment by MarcusSorealheis** [comment]
My old homie @jdbranham I'm going to checkout this PR and see if I can reproduce to fix the issue. It's such a great feature. 

Can you share a bit of info about your environment? I'm guessing you won't see this today or be at your computer but knowing you...

**Comment by jdbranham** [comment]
@MarcusSorealheis ! - I didn't know you were contributor. You're involved in all the cool stuff! =)

It's Friday night - what else is there to do  😅 
Thanks for taking a look.

```
❯ node -v
v16.14.2

❯ npm -v
8.5.0

❯ arch
arm64
```

**Comment by villebro** [comment]
@jdbranham I tried to open a PR agaist your fork, but it got really messy (probably due to GH choking on too many forks). But this commit should fix the licence header CI issues + fix the unit test: https://github.com/apache/superset/commit/bc3ee852666a59a53c50d16b2d291fc911bfa74d

**Review by villebro — APPROVED** [review]
LGTM, finally! 😄👍🚀

**Pull request merged**
